File: exporter_service.go
package main

import (
	"context"

	"github.com/rs/zerolog/log"
)

//go:generate mockgen -package=main -destination ./exporter_service_mock.go -source=exporter_service.go
type ExporterService interface {
	Run(ctx context.Context, bigqueryInit bool, bigqueryDataset, bigqueryTable string) (err error)
}

func NewExporterService(configClient ConfigClient, bigqueryClient BigQueryClient, stateClient StateClient, modbusClient ModbusClient) (ExporterService, error) {
	return &exporterService{
		configClient:   configClient,
		bigqueryClient: bigqueryClient,
		stateClient:    stateClient,
		modbusClient:   modbusClient,
	}, nil
}

type exporterService struct {
	configClient   ConfigClient
	bigqueryClient BigQueryClient
	stateClient    StateClient
	modbusClient   ModbusClient
}

func (s *exporterService) Run(ctx context.Context, bigqueryInit bool, bigqueryDataset, bigqueryTable string) (err error) {

	// read config from yaml file
	config, err := s.configClient.ReadConfigFromFile(ctx, *configPath)
	if err != nil {
		return
	}

	log.Info().Interface("config", config).Msgf("Loaded config from %v", *configPath)

	// init bigquery table if it doesn't exist yet
	if bigqueryInit {
		err = s.bigqueryClient.InitBigqueryTable(ctx, bigqueryDataset, bigqueryTable)
		if err != nil {
			return
		}
	}

	lastMeasurement, err := s.stateClient.ReadState(ctx)
	if err != nil {
		return
	}

	measurement, err := s.modbusClient.GetMeasurement(ctx, config, lastMeasurement)
	if err != nil {
		return
	}

	err = s.bigqueryClient.InsertMeasurement(ctx, bigqueryDataset, bigqueryTable, measurement)
	if err != nil {
		return
	}

	err = s.stateClient.StoreState(ctx, measurement)
	if err != nil {
		return
	}

	return nil
}

File: src/exporter_service.rs
use std::error::Error;

use crate::bigquery_client::BigqueryClient;
use crate::config_client::ConfigClient;
use crate::measurement_client::MeasurementClient;
use crate::state_client::StateClient;
use serde::de::DeserializeOwned;

pub struct ExporterServiceConfig<T: ?Sized> {
    config_client: ConfigClient,
    bigquery_client: BigqueryClient,
    state_client: StateClient,
    measurement_client: Box<dyn MeasurementClient<T>>,
}

impl<T> ExporterServiceConfig<T> {
    pub fn new(
        config_client: ConfigClient,
        bigquery_client: BigqueryClient,
        state_client: StateClient,
        measurement_client: Box<dyn MeasurementClient<T>>,
    ) -> Result<Self, Box<dyn Error>> {
        Ok(Self {
            config_client,
            bigquery_client,
            state_client,
            measurement_client,
        })
    }
}

pub struct ExporterService<T> {
    config: ExporterServiceConfig<T>,
}

impl<T> ExporterService<T> {
    pub fn new(config: ExporterServiceConfig<T>) -> Self {
        Self { config }
    }

    pub async fn run(&self) -> Result<(), Box<dyn std::error::Error>>
    where
        T: DeserializeOwned,
    {
        let config: T = self.config.config_client.read_config_from_file()?;

        self.config.bigquery_client.init_table().await?;

        let last_measurement = self.config.state_client.read_state()?;

        let measurement = self
            .config
            .measurement_client
            .get_measurement(config, last_measurement)?;

        self.config
            .bigquery_client
            .insert_measurement(&measurement)
            .await?;

        self.config.state_client.store_state(&measurement).await?;

        Ok(())
    }
}

// func TestRun(t *testing.T) {
// 	t.Run("ReadsConfigFromFile", func(t *testing.T) {

// 		ctx := context.Background()
// 		ctrl := gomock.NewController(t)
// 		defer ctrl.Finish()

// 		configClient := NewMockConfigClient(ctrl)
// 		bigqueryClient := NewMockBigQueryClient(ctrl)
// 		stateClient := NewMockStateClient(ctrl)
// 		modbusClient := NewMockModbusClient(ctrl)

// 		bigqueryInit := true
// 		bigqueryDataset := "dataset"
// 		bigqueryTable := "table"
// 		config := Config{}
// 		measurement := contractsv1.Measurement{}

// 		service, _ := NewExporterService(configClient, bigqueryClient, stateClient, modbusClient)

// 		configClient.EXPECT().ReadConfigFromFile(ctx, gomock.Any()).Return(config, nil)
// 		bigqueryClient.EXPECT().InitBigqueryTable(ctx, bigqueryDataset, bigqueryTable).Return(nil)
// 		stateClient.EXPECT().ReadState(ctx).Return(nil, nil)
// 		modbusClient.EXPECT().GetMeasurement(ctx, config, nil).Return(measurement, nil)
// 		bigqueryClient.EXPECT().InsertMeasurement(ctx, bigqueryDataset, bigqueryTable, measurement).Return(nil)
// 		stateClient.EXPECT().StoreState(ctx, measurement).Return(nil)

// 		// act
// 		err := service.Run(ctx, bigqueryInit, bigqueryDataset, bigqueryTable)

// 		assert.Nil(t, err)
// 	})
// }
